File: MatlabOctaveCodeSnippets/snip_frequency_msspectrum.m
N=1024; x = transpose(10*cos(2*3/64*(0:N-1))); %generate a cosine
Xk = (abs(fft(x,N))/N).^2; %MS spectrum: |DTFS|^2
Fs = 2*pi; myWindow = hamming(N); %specify Fs and a Hamming window
H = Fs*pwelch(x,myWindow,[],N,Fs,'twosided');%Welch's estimate. Third
%parameter is [] because in Matlab is num. samples while Octave is %
H = H * sum(myWindow.^2)/sum(myWindow)^2; %scale for MS
plot(Xk,'x-'), hold on, plot(H,'or-') %compare

File: PythonCodeSnippets/snip_frequency_msspectrum.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import welch
from scipy.signal.windows import hamming
from scipy.fft import fft

N = 1024  # Number of samples
x = 10 * np.cos(2 * 3 / 64 * np.arange(N))  # generate a cosine
Xk = (np.abs(fft(x)) / N) ** 2  # MS spectrum: |DTFS|^2
Fs = 2 * np.pi  # specify sampling frequence
myWindow = hamming(N)  # specify a Hamming window
_, H = welch(
    x, fs=Fs, window=myWindow, nfft=N, return_onesided=False
)  # Welch's estimate
H = Fs * H * np.sum(myWindow**2) / np.sum(myWindow) ** 2  # scale for MS
plt.plot(Xk, marker="X")  # compare
plt.plot(H, marker=".")
